packages/mui-utils/src/resolveProps.ts

```diff
@@ -11,7 +15,12 @@ export default function resolveProps<T extends { className?: string } & Record<s
   const output = { ...props };
 
   Object.keys(defaultProps).forEach((propName: keyof T) => {
-    if (output[propName] === undefined) {
+    const defaultValue = defaultProps[propName];
+    const currentValue = output[propName];
+
+    if (isObject(defaultValue) && isObject(currentValue)) {
+      output[propName] = { ...defaultValue, ...currentValue };
+    } else if (output[propName] === undefined) {
       output[propName] = defaultProps[propName];
     }
```

Comment:
_**Minor nit**_: I would keep component props check in the first conditional branch since _most of the time_ it would not be present if `defaultProps` are added in theme and also not many users complained about this issue. This would reduce execution time: 
```suggestion
    if (output[propName] === undefined) {
      output[propName] = defaultProps[propName];
    } else if (isObject(defaultValue) && isObject(currentValue)) {
      output[propName] = { ...defaultValue, ...currentValue };
    }
```